File: behavioral/dart/interpreter_structured/interfaces.dart
abstract class Expression {
  int interpret();
}

abstract class BooleanExpression {
  bool evaluate(Context context);
}

abstract class PatternExpression {
  bool match(String text);
}

class Context {
  final Map<String, int> _variables = {};
  
  void setVariable(String name, int value) {
    _variables[name] = value;
  }
  
  int getVariable(String name) {
    return _variables[name] ?? 0;
  }
}

File: behavioral/typescript/interpreter_structured/interfaces.ts
export interface Expression {
  interpret(): number;
}

export interface BooleanExpression {
  evaluate(context: Context): boolean;
}

export interface PatternExpression {
  match(text: string): boolean;
}

export class Context {
  private variables: Map<string, number> = new Map();
  
  setVariable(name: string, value: number): void {
    this.variables.set(name, value);
  }
  
  getVariable(name: string): number {
    return this.variables.get(name) || 0;
  }
}
